guide.md

```diff
@@ -0,0 +1,406 @@
+later).
+
+Now we're going to write a small CLI that lets us manage a blog (ignoring the
+fact that we can only access the database from this CLI...). The first thing
+we're going to need is a table to store our posts. Let's create a migration for
+that:
+
+```shell
+diesel migration generate create_posts_table
+```
+
+You should see output which looks something like this:
+
+```
+Creating migrations/20160202154039_create_posts/up.sql
+Creating migrations/20160202154039_create_posts/down.sql
+```
+
+Migrations allow us to evolve the database schema over time. Each migration can
+be applied (`up.sql`) or reverted (`down.sql`). Applying and immediately
+reverting a migration should leave your database schema unchanged.
+
+```sql
+-- up.sql
+CREATE TABLE posts (
+  id SERIAL PRIMARY KEY,
+  title VARCHAR NOT NULL,
+  body TEXT NOT NULL,
+  published BOOLEAN NOT NULL DEFAULT 'f'
+)
+
+-- down.sql
+DROP TABLE posts
+```
+
+We can apply our new migration with `diesel migration run`. It's a good idea to
+make sure that `down.sql` is correct. You can do a quick sanity check by doing
+`diesel migration redo`, which will revert and reapply the latest migration.
+
+OK enough SQL, let's write some Rust. We'll start by writing some code to show
+us the last 5 published posts. The first thing we need to do is establish a
+database connection.
+
+```rust
+// src/lib.rs
+#[macro_use]
+extern crate diesel;
+extern crate dotenv;
+
+use diesel::prelude::*;
+use diesel::pg::PgConnection;
+use dotenv::dotenv;
+use std::env;
+
+pub fn establish_connection() -> PgConnection {
+    dotenv().ok();
+
+    let database_url = env::var("DATABASE_URL")
+        .expect("DATABASE_URL must be set");
+    PgConnection::establish(&database_url)
+        .expect(&format!("Error connecting to {}", database_url))
+}
+```
+
+We'll also want to create a `Post` struct that we can read our data into, and
+have diesel generate the names we'll use to reference our tables and columns in
+our queries.
+
+We'll add the following two lines to the top of `src/lib.rs`
+
+```rust
+#![feature(custom_derive, custom_attribute, plugin)]
+#![plugin(diesel_codegen, dotenv_macros)]
+
+pub mod schema;
+pub mod models;
+```
+
+The first two lines tell Rust that we want to use some special compiler plugins
+provided by Diesel and Dotenv. These will add various useful class annotations
+that we can use, as well as the `infer_schema!` macro which we'll see in just a
+moment. Next we need to create the two modules that we just declared.
+
+```rust
+// src/models.rs
+#[derive(Queryable)]
+pub struct Post {
+    pub id: i32,
+    pub title: String,
+    pub body: String,
+    pub published: bool,
+}
+
+// src/schema.rs
+infer_schema!(dotenv!("DATABASE_URL"));
+```
+
+The `#[derive(Queryable)]` will generate all of the code needed to load a `Post`
+struct from a SQL query. The `infer_schema!` macro connects to the database URL
+given to it, and creates a bunch of code based on the database schema to
```

Comment:
"The `infer_schema!` macro connects to the database URL which it is given. The macro also creates a bunch of code based on the database schema to represent all of the tables and columns."